js/bruh-dash.js

```diff
@@ -127,14 +285,49 @@ global.bruhdash = {
 
   // iterates over elements of a collection and returns all elements that the predicate returns truthy for
   // Note: this should work for arrays and objects
-  filter: function() {
-
+  //returns an array of filtered elements - all the elements that passed the verification condition/function
+
+  filter: function(col, func) {
+    var newEl;
+    var newArr = [];
+    if (Array.isArray(col)){
+      for (var i = 0; i < col.length; i++){
+        if (func(col[i])){
+          newEl = col[i];
+          newArr.push(newEl);
+        }
+      }
+    } else if(typeof col === 'object'){
+        for (var key in col){
+          if (func(col[key]) === true){
+            newEl = col[key];
+            newArr.push(newEl);
+          }
+        }
+    }
+    return newArr;
+    //is this an array of objects or properties of a single object?
+    //what is it returning? - an array? 
+    //the intention of this function when it comes to objects isn't clear to me
   },
 
   // Reduces the collection to a value which is the accumulated result of running each element
   // in the collection through an iteratee
   // Note: this should work for arrays and objects
-  reduce: function() {
-    
+
+  reduce: function(col, func, initial) {
+    var result = 0;
+
+    if (Array.isArray(col)){
+      for (var i = 0; i < col.length; i++){
+        result = func(result,col[i]);
+      }
+    } else if(typeof col === 'object'){
+      for (var key in col){
+        result = func(result, col[key]);
+      }
+    }
+    return result;
   }
+  //the way the test is applied to this makes zero sense to me - why would we want to sum up the values for the different keys? Is that really how we would iterate through the object? Or is this method meant for iterating through objects (rather than their keys) and summing up values of the same key in different objects?
```

Comment:
So the point of a reduce function is to iterate through an array or object, apply a function at each index (or value for object) then keep track of the sum of each function output. So the classical example would be, if you had an object where each key is paired with a value that is a sentence. If we wanted to get a word count for all the sentences in the object, a reduce function makes this simple. It's even more useful if there's logic that needs to be done in the function being applied to each value as well. Such as getting a sum of all the nouns across all sentences in the object.